tiri/src/main.rs

```diff
@@ -0,0 +1,121 @@
+}
+
+impl<'p> Frame<'p> {
+    /// Create a new frame and position its program counter at the beginning of the first block.
+    fn new(body: &'p Body) -> Self {
+        Self { body, pc: (0, 0) }
+    }
+}
+
+/// The interpreter itself.
+/// The struct itself holds only immutable program information.
+struct Interp {
+    /// Maps a `DefId` to the corresponding TIR body.
+    tir_map: HashMap<DefId, Body>,
+    /// The `DefId` of the `main()` function, which serves as the entry point.
+    main_defid: DefId,
+}
+
+impl Interp {
+    /// Create a new interpreter, using the TIR found in the `.yk_tir` section of the binary `bin`.
+    fn new(bin: &Path) -> Self {
+        let ef = elf::File::open_path(&PathBuf::from(bin)).unwrap();
+        let sec = ef.get_section(".yk_tir").unwrap();
+        let mut curs = Cursor::new(&sec.data);
+        let mut dec = Decoder::from(&mut curs);
+
+        let mut tir_map = HashMap::new();
+        let mut main_defid = None;
+        while let Some(pack) = dec.next().unwrap() {
+            let Pack::Body(body) = pack;
+
+            if body.def_path_str == "main" {
+                main_defid = Some(body.def_id.clone());
+            }
+            tir_map.insert(body.def_id.clone(), body);
+        }
+
+        Self {
+            tir_map: tir_map,
+            main_defid: main_defid.expect("Couldn't find main()"),
+        }
+    }
+
+    /// Start interpreting TIR from the `main()` function.
+    fn run(&self) {
+        // Set up the stack to start executing at the entry point.
+        let mut stack = Vec::new();
+        stack.push(Frame::new(&self.tir_map[&self.main_defid]));
+
+        // The main interpreter loop.
+        loop {
+            let cur_frame = stack.last().unwrap();
+            let cur_block = &cur_frame.body.blocks[usize::try_from(cur_frame.pc.0).unwrap()];
+            let block_len = cur_block.stmts.len();
+            let pc_stmt_usize = usize::try_from(cur_frame.pc.1).unwrap();
+
+            if pc_stmt_usize < block_len {
+                let stmt = &cur_block.stmts[pc_stmt_usize];
+                debug!(
+                    "Interp Loc: body={}, pc={:?}:\n  {}",
+                    cur_frame.body.def_path_str, cur_frame.pc, stmt
+                );
+                self.interp_stmt(stmt);
+            } else if pc_stmt_usize == block_len {
+                // We take statement index one past the end to mean the block terminator.
+                let term = &cur_block.term;
+                debug!(
+                    "Interp Term: body={}, pc={:?}:\n  {}",
+                    cur_frame.body.def_path_str, cur_frame.pc, term
+                );
+                self.interp_term(term);
+            } else {
+                unreachable!();
+            }
+
+            break; // FIXME -- Just to prevent infinite loop until statement actions are implemented.
+        }
+    }
+
+    /// Interpret the specified statement.
+    fn interp_stmt(&self, _stmt: &Statement) {}
```

Comment:
Should have `unimplemented!()` in the body?

Reply:
We could, but I wouldn't worry too much, as these are the next items to be populated.

Reply:
I'm a big fan of `unimplemented`, because it's also a clear guide to other people (e.g. reviewers) that things are as yet unfilled.